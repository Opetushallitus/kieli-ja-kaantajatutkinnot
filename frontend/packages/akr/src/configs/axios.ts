import axios from 'axios';

import { AppConstants } from 'enums/app';

const axiosInstance = axios.create();

axiosInstance.interceptors.request.use((config) => {
  const CSRF_NAME = 'CSRF';

  config.xsrfCookieName = CSRF_NAME;
  config.xsrfHeaderName = CSRF_NAME;
  config.withCredentials = true;

  config.headers = {
    ...config.headers,
    'Caller-Id': AppConstants.CallerID,
    'Content-Type': 'application/json',
  };

  return config;
});

export default axiosInstance;
