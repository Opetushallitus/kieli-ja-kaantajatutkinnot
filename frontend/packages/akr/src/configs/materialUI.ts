import { createTheme } from '@mui/material/styles';

// Create Material UI theme configs
const primaryColor = '#FFFFFF';
const primaryLightColor = '#F5F5F5';
const primaryDarkColor = '#CCCCCC';
const secondaryColor = '#0041dc';
const secondaryLightColor = '#159ECB';
const secondaryDarkColor = '#000a48';
const primaryHeadingColor = '#000a48';
const secondaryHeadingColor = '#FFFFFF';
const grey700Color = '#666666';
const grey600Color = '#999999';
const blue200Color = '#F2F5FD';
const red500Color = '#db2828';

const fontWeightBold = 700;
const fontWeightMedium = 500;
const fontWeightRegular = 400;

export const theme = createTheme({
  components: {
    MuiOutlinedInput: {
      styleOverrides: {
        root: {
          '&.Mui-focused .MuiOutlinedInput-notchedOutline': {
            borderColor: secondaryDarkColor,
          },
          color: grey700Color,
          '&.Mui-disabled': {
            backgroundColor: primaryLightColor,
          },
        },
      },
    },
    MuiFormLabel: {
      styleOverrides: {
        root: {
          '&.Mui-focused': {
            color: grey700Color,
          },
          '&.Mui-disabled': {
            color: primaryHeadingColor,
          },
          color: grey700Color,
        },
      },
    },
    MuiCheckbox: {
      styleOverrides: {
        root: {
          color: grey700Color,
        },
      },
    },
    MuiStepIcon: {
      styleOverrides: {
        root: {
          '&.Mui-completed': {
            color: secondaryColor,
          },
          '&.Mui-active': {
            color: secondaryColor,
          },
        },
        text: {
          fill: primaryColor,
        },
      },
    },
    MuiStepLabel: {
      styleOverrides: {
        label: {
          '&.Mui-disabled': {
            color: grey600Color,
          },
        },
      },
    },
    MuiFormHelperText: {
      styleOverrides: {
        root: {
          color: grey700Color,
        },
      },
    },
    MuiTab: {
      styleOverrides: {
        root: {
          color: grey700Color,
        },
      },
    },
    MuiSkeleton: {
      styleOverrides: {
        root: {
          backgroundColor: blue200Color,
        },
      },
    },
    MuiButton: {
      styleOverrides: {
        root: {
          '&.Mui-disabled': {
            cursor: 'not-allowed',
            pointerEvents: 'auto',
          },
        },
      },
    },
    MuiTablePagination: {
      styleOverrides: {
        menuItem: {
          '&.Mui-selected': {
            backgroundColor: primaryLightColor,
          },
        },
      },
    },
    MuiTooltip: {
      styleOverrides: {
        tooltip: {
          fontSize: '1.4rem',
        },
      },
    },
  },
  palette: {
    error: {
      main: red500Color,
    },
    primary: {
      main: primaryColor,
      light: primaryLightColor,
      dark: primaryDarkColor,
    },
    secondary: {
      main: secondaryColor,
      light: secondaryLightColor,
      dark: secondaryDarkColor,
    },
    text: {
      primary: primaryHeadingColor,
      secondary: secondaryHeadingColor,
    },
  },
  typography: {
    htmlFontSize: 10,
    h1: {
      fontSize: '2.8rem',
      fontWeight: fontWeightMedium,
      lineHeight: '3.3rem',
      marginBottom: '2rem',
      color: primaryHeadingColor,
    },
    h2: {
      fontSize: '2rem',
      fontWeight: fontWeightMedium,
      lineHeight: '2.4rem',
      color: primaryHeadingColor,
    },
    h3: {
      fontSize: '1.6rem',
      fontWeight: fontWeightBold,
      lineHeight: '1.9rem',
      color: primaryHeadingColor,
    },
    body1: {
      fontSize: '1.6rem',
      fontWeight: fontWeightRegular,
      lineHeight: '2.4rem',
      color: primaryHeadingColor,
    },
    caption: {
      fontSize: '1.2rem',
      fontWeight: fontWeightRegular,
      color: grey700Color,
    },
  },
});
