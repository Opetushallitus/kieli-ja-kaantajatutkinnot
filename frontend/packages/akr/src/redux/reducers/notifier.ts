import { Reducer } from 'redux';

import {
  Dialog,
  NotifierAction,
  NotifierState,
  Toast,
} from 'interfaces/notifier';
import {
  NOTIFIER_DIALOG_ADD,
  NOTIFIER_DIALOG_REMOVE,
  NOTIFIER_TOAST_ADD,
  NOTIFIER_TOAST_REMOVE,
} from 'redux/actionTypes/notifier';

const defaultState = {
  dialogs: [],
  toasts: [],
};

export const notifierReducer: Reducer<NotifierState, NotifierAction> = (
  state = defaultState,
  action
) => {
  switch (action.type) {
    case NOTIFIER_DIALOG_ADD:
      return {
        ...state,
        dialogs: [...state.dialogs, <Dialog>action.notifier],
      };
    case NOTIFIER_DIALOG_REMOVE:
      return {
        ...state,
        dialogs: state.dialogs.filter((d) => d.id != action.id),
      };
    case NOTIFIER_TOAST_ADD:
      return {
        ...state,
        toasts: [...state.toasts, <Toast>action.notifier],
      };
    case NOTIFIER_TOAST_REMOVE:
      return {
        ...state,
        toasts: state.toasts.filter((d) => d.id != action.id),
      };
    default:
      return state;
  }
};
