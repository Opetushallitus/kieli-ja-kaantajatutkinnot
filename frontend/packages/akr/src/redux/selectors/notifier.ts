import { RootState } from 'configs/redux';

export const notificationSelector = (state: RootState) => state.notifier;
