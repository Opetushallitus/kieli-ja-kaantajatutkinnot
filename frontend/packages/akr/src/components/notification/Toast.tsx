import { Alert, Snackbar } from '@mui/material';
import { useEffect, useState } from 'react';

import { useAppDispatch, useAppSelector } from 'configs/redux';
import { Notifier } from 'interfaces/notifier';
import { removeNotifierToast } from 'redux/actions/notifier';
import { notificationSelector } from 'redux/selectors/notifier';

export const Toast = () => {
  // Redux
  const dispatch = useAppDispatch();
  const { toasts } = useAppSelector(notificationSelector);

  // State
  const [activeToast, setActiveToast] = useState<Notifier | undefined>(
    undefined
  );

  const handleToastClose = (id: string) => {
    dispatch(removeNotifierToast(id));
    setActiveToast(undefined);
  };

  useEffect(() => {
    if (toasts?.length > 0 && !activeToast) {
      const [toast] = toasts;
      setActiveToast(toast);
    }
  }, [activeToast, toasts]);

  return (
    <>
      {activeToast && (
        <Snackbar
          anchorOrigin={{ vertical: 'top', horizontal: 'left' }}
          open={!!activeToast}
          autoHideDuration={activeToast.timeOut}
          onClose={() => handleToastClose(activeToast.id)}
        >
          <Alert
            data-testid="toast-notification"
            variant={'filled'}
            onClose={() => handleToastClose(activeToast.id)}
            severity={activeToast.severity}
          >
            {activeToast.description}
          </Alert>
        </Snackbar>
      )}
    </>
  );
};
