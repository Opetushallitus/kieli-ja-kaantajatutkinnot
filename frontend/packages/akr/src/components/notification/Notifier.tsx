import { DialogBox } from 'components/notification/DialogBox';
import { Toast } from 'components/notification/Toast';

export const Notifier = () => (
  <>
    <DialogBox />
    <Toast />
  </>
);
