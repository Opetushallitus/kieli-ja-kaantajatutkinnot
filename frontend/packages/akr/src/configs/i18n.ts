import i18n, { changeLanguage, t, use } from 'i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
import {
  initReactI18next,
  useTranslation,
  UseTranslationOptions,
} from 'react-i18next';
import { AppLanguage, I18nNamespace } from 'shared/enums';
import { DateUtils } from 'shared/utils';

import accessibilityEN from 'public/i18n/en-GB/accessibility.json';
import commonEN from 'public/i18n/en-GB/common.json';
import privacyEN from 'public/i18n/en-GB/privacy.json';
import transEN from 'public/i18n/en-GB/translation.json';
import accessibilityFI from 'public/i18n/fi-FI/accessibility.json';
import commonFI from 'public/i18n/fi-FI/common.json';
import privacyFI from 'public/i18n/fi-FI/privacy.json';
import transFI from 'public/i18n/fi-FI/translation.json';
import koodistoCountriesEN from 'public/i18n/koodisto/countries/koodisto_countries_en-GB.json';
import koodistoCountriesFI from 'public/i18n/koodisto/countries/koodisto_countries_fi-FI.json';
import koodistoCountriesSV from 'public/i18n/koodisto/countries/koodisto_countries_sv-SE.json';
import koodistoLangsEN from 'public/i18n/koodisto/langs/koodisto_langs_en-GB.json';
import koodistoLangsFI from 'public/i18n/koodisto/langs/koodisto_langs_fi-FI.json';
import koodistoLangsSV from 'public/i18n/koodisto/langs/koodisto_langs_sv-SE.json';
import accessibilitySV from 'public/i18n/sv-SE/accessibility.json';
import commonSV from 'public/i18n/sv-SE/common.json';
import privacySV from 'public/i18n/sv-SE/privacy.json';
import transSV from 'public/i18n/sv-SE/translation.json';

// Defaults and resources
const langFI = AppLanguage.Finnish;
const langSV = AppLanguage.Swedish;
const langEN = AppLanguage.English;

const supportedLangs = [langFI, langSV, langEN];

const resources = {
  [langFI]: {
    [I18nNamespace.Accessibility]: accessibilityFI,
    [I18nNamespace.Common]: commonFI,
    [I18nNamespace.KoodistoCountries]: koodistoCountriesFI,
    [I18nNamespace.KoodistoLanguages]: koodistoLangsFI,
    [I18nNamespace.Privacy]: privacyFI,
    [I18nNamespace.Translation]: transFI,
  },
  [langSV]: {
    [I18nNamespace.Accessibility]: accessibilitySV,
    [I18nNamespace.Common]: commonSV,
    [I18nNamespace.KoodistoCountries]: koodistoCountriesSV,
    [I18nNamespace.KoodistoLanguages]: koodistoLangsSV,
    [I18nNamespace.Privacy]: privacySV,
    [I18nNamespace.Translation]: transSV,
  },
  [langEN]: {
    [I18nNamespace.Accessibility]: accessibilityEN,
    [I18nNamespace.Common]: commonEN,
    [I18nNamespace.KoodistoCountries]: koodistoCountriesEN,
    [I18nNamespace.KoodistoLanguages]: koodistoLangsEN,
    [I18nNamespace.Privacy]: privacyEN,
    [I18nNamespace.Translation]: transEN,
  },
};

const detectionOptions = {
  order: ['localStorage', 'htmlTag'],
  caches: ['localStorage'],
  lookupLocalStorage: 'i18NextLng_akr',
};

// TypeScript definitions for react-i18next. IDE might show this to be unused, but ts does some type checking with it.
declare module 'react-i18next' {
  interface CustomTypeOptions {
    defaultNS: typeof langFI;
    resources: {
      [langFI]: typeof transFI;
      [langSV]: typeof transSV;
      [langEN]: typeof transEN;
    };
  }
}

export const initI18n = () => {
  const i18n = use(initReactI18next).use(LanguageDetector).init({
    resources,
    detection: detectionOptions,
    fallbackLng: langFI,
    load: 'currentOnly',
    debug: !REACT_ENV_PRODUCTION,
  });
  const currentLanguage = getCurrentLang() as AppLanguage;
  DateUtils.setDayjsLocale(currentLanguage);

  return i18n;
};

export const useAppTranslation = (
  options: UseTranslationOptions<string>,
  ns: I18nNamespace = I18nNamespace.Translation
) => {
  // @ts-expect-error ts import fail
  return useTranslation(ns, options);
};

export const useAccessibilityTranslation = () => {
  const { t } = useAppTranslation(
    {
      keyPrefix: 'akr.accessibility',
    },
    I18nNamespace.Accessibility
  );

  return t;
};

export const useCommonTranslation = () => {
  const { t } = useAppTranslation(
    {
      keyPrefix: 'akr.common',
    },
    I18nNamespace.Common
  );

  return t;
};

export const useKoodistoCountriesTranslation = () => {
  const { t } = useAppTranslation(
    {
      keyPrefix: 'akr.koodisto.countries',
    },
    I18nNamespace.KoodistoCountries
  );

  return t;
};

export const useKoodistoLanguagesTranslation = () => {
  const { t } = useAppTranslation(
    {
      keyPrefix: 'akr.koodisto.languages',
    },
    I18nNamespace.KoodistoLanguages
  );

  return t;
};

export const usePrivacyTranslation = () => {
  const { t } = useAppTranslation(
    {
      keyPrefix: 'akr.privacy',
    },
    I18nNamespace.Privacy
  );

  return t;
};

export const translateOutsideComponent = () => {
  return t;
};

export const getCurrentLang = (): AppLanguage => {
  return i18n.language as AppLanguage;
};

export const isCurrentLangSv = (): boolean => {
  return getCurrentLang() === AppLanguage.Swedish;
};

export const getSupportedLangs = (): Array<AppLanguage> => {
  return supportedLangs;
};

export const changeLang = (language: AppLanguage) => {
  DateUtils.setDayjsLocale(language);

  return changeLanguage(language);
};
