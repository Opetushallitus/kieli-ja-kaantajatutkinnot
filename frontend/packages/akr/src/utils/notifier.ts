import { AxiosError } from 'axios';

import { translateOutsideComponent } from 'configs/i18n';
import { APIError } from 'enums/api';

export class NotifierUtils {
  static getAPIErrorMessage(error: AxiosError) {
    const t = translateOutsideComponent();
    const apiError = NotifierUtils.getAPIError(error);

    const message = apiError
      ? t(`akr.errors.api.${apiError}`)
      : t('akr.errors.api.generic');

    return message;
  }

  private static getAPIError(error: AxiosError) {
    const errorCode = error.response?.data.errorCode;

    if (errorCode && Object.values(APIError).includes(errorCode)) {
      return errorCode;
    }
  }
}
