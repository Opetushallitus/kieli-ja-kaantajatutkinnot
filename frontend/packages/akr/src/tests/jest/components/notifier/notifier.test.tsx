import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { DialogBox } from 'components/notification/DialogBox';
import { Toast } from 'components/notification/Toast';
import { useAppSelector } from 'configs/redux';
import { NotifierState } from 'interfaces/notifier';
import {
  removeNotifierDialog,
  removeNotifierToast,
  showNotifierDialog,
  showNotifierToast,
} from 'redux/actions/notifier';
import { notifierReducer } from 'redux/reducers/notifier';
import {
  dialogsArray,
  emptyNotifierState,
  notifierState,
  toastsArray,
} from 'tests/jest/__fixtures__/notifier';

describe('DialogBox', () => {
  it('should not render Dialog if there is no data', () => {
    const { container } = render(<DialogBox />);

    expect(container).toBeEmptyDOMElement();
  });

  it('should render DialogBox correctly', () => {
    initializeState(notifierState);
    const { container } = render(<DialogBox />);

    expect(container).toMatchSnapshot();
  });

  it('should show a Dialog when an action is dispatched', () => {
    const [dialog] = dialogsArray;
    const previousState = emptyNotifierState;
    const dialogCreator = showNotifierDialog(dialog);

    const newState = notifierReducer(previousState, dialogCreator);

    expect(newState).toEqual({ dialogs: [dialog], toasts: [] });
  });

  it('should remove a Dialog when an action is dispatched', () => {
    const [dialog] = dialogsArray;
    const previousState = { dialogs: [dialog], toasts: [] };
    const dialogCreator = removeNotifierDialog(dialog.id);

    const newState = notifierReducer(previousState, dialogCreator);

    expect(newState).toEqual({ dialogs: [], toasts: [] });
  });
});

describe('Toast', () => {
  it('should not render Toast if there is no data', () => {
    initializeState(emptyNotifierState);
    const { container } = render(<Toast />);

    expect(container).toBeEmptyDOMElement();
  });

  it('should render Toast correctly', () => {
    initializeState(notifierState);
    const tree = renderer.create(<Toast />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should show a Toast when an action is dispatched', () => {
    const [toast] = toastsArray;
    const previousState = emptyNotifierState;
    const toastCreator = showNotifierToast(toast);

    const newState = notifierReducer(previousState, toastCreator);

    expect(newState).toEqual({ dialogs: [], toasts: [toast] });
  });

  it('should remove a Toast when an action is dispatched', () => {
    const [toast] = toastsArray;
    const previousState = { dialogs: [], toasts: [toast] };
    const toastCreator = removeNotifierToast(toast.id);

    const newState = notifierReducer(previousState, toastCreator);

    expect(newState).toEqual({ dialogs: [], toasts: [] });
  });
});

// Helpers and Mocks
const initializeState = (state: NotifierState) => {
  const appSelector = useAppSelector as jest.Mock<NotifierState>;
  appSelector.mockImplementation(() => state);
};
