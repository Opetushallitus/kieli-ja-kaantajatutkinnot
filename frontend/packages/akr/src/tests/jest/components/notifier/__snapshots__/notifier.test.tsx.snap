// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogBox should render DialogBox correctly 1`] = `
<div
  aria-hidden="true"
/>
`;

exports[`Toast should render Toast correctly 1`] = `null`;
